<!DOCTYPE html>
<head>
<meta charset="utf-8">

<title>Burrito Visualization</title>

<script src="shared/jquery.js" type="text/javascript"></script>
<script src="shared/shiny.js" type="text/javascript"></script>
<link rel="stylesheet" type="text/css" href="shared/shiny.css"/>
<script src="__assets__/shiny-server.js"></script>
<link rel="stylesheet" type="text/css" href="CSS/burritostyle.css"/>

</head>

<body onresize="resizeRedraw()">
<div id="mainthing" style="overflow-x:auto">


<img class="menuEl" border="0" src="images/title_lab_800.jpg"></img>

<h2>Burrito: A Marriage of Taxonomy and Function</h2> 
<img class="menuEl" border="1" src="images/bEXnotext.jpg" width="400"></img>		

<p>This is a tool to visualize the relationship between taxonomy and function. You can view sample data or upload your own data.</p>
<p>Source code available <a href="https://github.com/borenstein-lab/burrito">here</a></p>
<p>Preview the tool with an example data set: <button type="button" name="example_button" id="example_button">Taste the burrito</button></p>
<p>Or upload your own data, choosing one of the three upload options below:<p>

<button class="inputSelectorB" id="ReadOption">
	<b>OTU table</b>
	<br/><br/>
	Choose this option if you have an OTU table from 16S reads, like from QIIME.
</button>
<div id="ReadMenu" class="hidexp">

<table class="uploadtable" width=100%>
	<col>
	<col>
	<col align="right">
	<tr>
		<td colspan=3>
			<b>File Upload</b>
		</td>
	</tr>
	<tr>
		<td>
			<div id="read_counts_g">
				<b>16S rRNA OTU Table<b>
				<br>
				<input type="file" name="read_counts" id="read_counts">
			</div>
		</td>
		<td>
			A file containing abundances of each taxon in each sample. See documentation for an example.
		</td>
		<td>
			<table class="ExampleTable">
				<tr>
					<th>OTU</th>
					<th>Sample1_Name</th>
					<th>Sample2_Name</th>
					<th>etc.</th>
				</tr>
				<tr>
					<td>OTU1_ID</td>
					<td>10</td>
					<td>5</td>
					<td>...</td>
				</tr>
				<tr>
					<td>OTU2_ID</td>
					<td>0</td>
					<td>5</td>
					<td>...</td>
				</tr>
				<tr>
					<td>etc.</td>
					<td>...</td>
					<td>...</td>
					<td>...</td>
				</tr>
			</table>
		</td>
	</tr>
	<tr>
		<td>
			Taxonomic Hierarchy
			<input type="file" name="taxonomic_hierarchy_R" id="taxonomic_hierarchy_R">
		</td>
		<td>
			A file containing all taxa in your abundance file, in a hierarchy showing their relatedness. See documentation for a format example. Uses greengenes by default if no file uploaded.
		</td>
		<td>
			<table class="ExampleTable">
				<tr>
					<th>OTU</th>
					<th>Taxa_Level_1</th>
					<th>Taxa_Level_2</th>
					<th>etc.</th>
				</tr>
				<tr>
					<td>OTU1_ID</td>
					<td>Bacteria</td>
					<td>Firmicutes</td>
					<td>...</td>
				</tr>
				<tr>
					<td>OTU2_ID</td>
					<td>Bacteria</td>
					<td>Proteobacteria</td>
					<td>...</td>
				</tr>
				<tr>
					<td>etc.</td>
					<td>...</td>
					<td>...</td>
					<td>...</td>
				</tr>
			</table>
		</td>
	</tr>
	<tr>
		<td>
			Function Hierarchy
			<input type="file" name="function_hierarchy_R" id="function_hierarchy_R">
		</td>
		<td>
			A file containing the functions in your contribution or genome content file, in a hierarchy showing their relatedness. See documentation for a format example. Uses BRITE by default if no file is uploaded.
		</td>
		<td>
			<table class="ExampleTable">
				<tr>
					<th>Gene</th>
					<th>Level_0</th>
					<th>Level_1</th>
					<th>etc.</th>
				</tr>
				<tr>
					<td>Gene1_Name</td>
					<td>Metabolism</td>
					<td>Lipid Metabolism</td>
					<td>...</td>
				</tr>
				<tr>
					<td>Gene2_Name</td>
					<td>Cellular Processes</td>
					<td>Cell Motility</td>
					<td>...</td>
				</tr>
				<tr>
					<td>etc.</td>
					<td>...</td>
					<td>...</td>
					<td>...</td>
				</tr>
			</table>
		</td>
	</tr>
	<tr>
		<td>
			Sample Map
			<input type="file" name="sample_map_R" id="sample_map_R">
		</td>
		<td>
			A file describing the relationship between samples, i.e. which ones are the same treatment condition.
		</td>
		<td>
			<table class="ExampleTable">
				<tr>
					<th>Sample</th>
					<th>Grouping_1</th>
					<th>Grouping_2</th>
					<th>etc.</th>
				</tr>
				<tr>
					<td>Sample1_Name</td>
					<td>Case</td>
					<td>Human</td>
					<td>...</td>
				</tr>
				<tr>
					<td>Sample2_Name</td>
					<td>Control</td>
					<td>Mouse</td>
					<td>...</td>
				</tr>
				<tr>
					<td>etc.</td>
					<td>...</td>
					<td>...</td>
					<td>...</td>
				</tr>
			</table>
		</td>
	</tr>
	<tr>
		<td>
			Function Abundances
			<input type="file" name="function_abundances_R" id="function_abundances_R">
		</td>
		<td>
			A file containing measured abundances of each function in each sample for comparison to function abundances predicted from taxonomic abundances.
		</td>
		<td>
			<table class="ExampleTable">
				<tr>
					<th>Gene</th>
					<th>Sample1_Name</th>
					<th>Sample2_Name</th>
					<th>etc.</th>
				</tr>
				<tr>
					<td>Gene1_Name</td>
					<td>0.1</td>
					<td>0.2</td>
					<td>...</td>
				</tr>
				<tr>
					<td>Gene2_Name</td>
					<td>0.3</td>
					<td>0.25</td>
					<td>...</td>
				</tr>
				<tr>
					<td>etc.</td>
					<td>...</td>
					<td>...</td>
					<td>...</td>
				</tr>
			</table>
		</td>
	</tr>

	<tr>
		<td colspan=3>
			<b>Visualization Options</b>
		</td>
	</tr>
	<tr>
		<td>
			Taxonomic level of detail
			<br>
			<select name="taxlevelofdetail_R" id="taxLODselector_R"></select>
		</td>
		<td colspan=2>
			Select the lowest level of taxonomic detail to be displayed. Going to the OTU level will significantly slow down the program.
		</td>
	</tr>
	<tr>
		<td>
			Function level of detail
			<br>
			<select name="funclevelofdetail_R" id="funcLODselector_R"></select>
		</td>
		<td colspan=2>
			Select the lowest level of function detail to be displayed. Going to the KO level is not recommended as single KOs can be mapped to multiple higher levels of function detail in the default hierarchy.
		</td>
	</tr>
	<tr>
		<td>
			Sample grouping
			<br>
			<select name="sampgroup_R" id="sampgroupselector_R"></select>
		</td>
		<td colspan=2>
			Select a grouping for samples if desired. Requires uploading a sample map.
		</td>
	</tr>
	<tr>
		<td>
			Color scale
			<br>
			<select name="color_option_selector_R" id="color_option_selector_R">
			<option value="Categories">Hierarchical color scale</option>
			<option value="Random">Random color scale</option>
			</select>
		</td>
		<td colspan=2>
			Color scheme. Random color scale will help discern closely related taxa but hierarchical color scale will look more pleasing.
		</td>
	</tr>
	<tr>
		<td colspan=3>
			<h4>Draw the burrito</h4>
		</td>
	</tr>
	<tr>
		<td colspan=3>
			<button id="update_button_R" type="button" class="btn action-button btn-large btn-primary">Assemble the Burrito</button>
		</td>
	</tr>
</table>

</div>

<button class="inputSelectorB" id="ContributionOption">
	<b>Functional contributions</b>
	<br/><br/>
	Choose this option if you have an OTU table from 16S reads and have run PICRUST yourself.
</button>

<div id="ContributionMenu" class="hidexp">
<table class="uploadtable" width=100%>
	<col>
	<col>
	<col align="right">
	<tr>
		<td colspan=3>
			<b>File Upload</b>
		</td>
	</tr>
	<tr>
		<td>
			<div id="tax_counts_g">
				<b>16S rRNA OTU Table<b>
				<br>
				<input type="file" name="taxonomic_abundances_2" id="taxonomic_abundances_2">
			</div>
		</td>
		<td>
			A file containing abundances of each taxon in each sample. See documentation for an example.
		</td>
		<td>
			<table class="ExampleTable">
				<tr>
					<th>OTU</th>
					<th>Sample1_Name</th>
					<th>Sample2_Name</th>
					<th>etc.</th>
				</tr>
				<tr>
					<td>OTU1_ID</td>
					<td>10</td>
					<td>5</td>
					<td>...</td>
				</tr>
				<tr>
					<td>OTU2_ID</td>
					<td>0</td>
					<td>5</td>
					<td>...</td>
				</tr>
				<tr>
					<td>etc.</td>
					<td>...</td>
					<td>...</td>
					<td>...</td>
				</tr>
			</table>
		</td>
	</tr>
	<tr>
		<td>
			<div id="func_contribution_g">
				<b>Function Contributions<b>
				<br>
				<input type="file" name="function_contributions" id="function_contributions">
			</div>
		</td>
		<td>
			An output file from PiCRUST, containing the contribution each taxa makes to each function.
		</td>
		<td>
			<div id="tooLongTableDiv">
				<div id="tooLongTable">
					<table class="ExampleTable">
						<tr>
							<th>Gene</th>
							<th>Sample</th>
							<th>OTU</th>
							<th>GeneCountPerGenome</th>
							<th>OTUAbundanceInSample</th>
							<th>CountContributedByOTU</th>
							<th>ContributionPercentOfSample</th>
							<th>ContributionPercentOfAllSamples</th>
						</tr>
						<tr>
							<td>Gene1_Name</td>
							<td>Sample1_Name</td>
							<td>OTU1_ID</td>
							<td>1.0</td>
							<td>10.0</td>
							<td>10.0</td>
							<td>0.5</td>
							<td>0.0002</td>
						</tr>
						<tr>
							<td>Gene1_Name</td>
							<td>Sample1_Name</td>
							<td>OTU2_ID</td>
							<td>2.0</td>
							<td>5.0</td>
							<td>10.0</td>
							<td>0.5</td>
							<td>0.0002</td>
						</tr>
						<tr>
							<td>etc.</td>
							<td>etc.</td>
							<td>etc.</td>
							<td>etc.</td>
							<td>etc.</td>
							<td>etc.</td>
							<td>etc.</td>
							<td>etc.</td>
						</tr>
					</table>
				</div>
			</div>
		</td>
	</tr>
	
	
	
	<tr>
		<td>
			Taxonomic Hierarchy
			<input type="file" name="taxonomic_hierarchy_C" id="taxonomic_hierarchy_C">
		</td>
		<td>
			A file containing all taxa in your abundance file, in a hierarchy showing their relatedness. See documentation for a format example. Uses greengenes by default if no file uploaded.
		</td>
		<td>
			<table class="ExampleTable">
				<tr>
					<th>OTU</th>
					<th>Taxa_Level_1</th>
					<th>Taxa_Level_2</th>
					<th>etc.</th>
				</tr>
				<tr>
					<td>OTU1_ID</td>
					<td>Bacteria</td>
					<td>Firmicutes</td>
					<td>...</td>
				</tr>
				<tr>
					<td>OTU2_ID</td>
					<td>Bacteria</td>
					<td>Proteobacteria</td>
					<td>...</td>
				</tr>
				<tr>
					<td>etc.</td>
					<td>...</td>
					<td>...</td>
					<td>...</td>
				</tr>
			</table>
		</td>
	</tr>
	<tr>
		<td>
			Function Hierarchy
			<input type="file" name="function_hierarchy_C" id="function_hierarchy_C">
		</td>
		<td>
			A file containing the functions in your contribution or genome content file, in a hierarchy showing their relatedness. See documentation for a format example. Uses BRITE by default if no file is uploaded.
		</td>
		<td>
			<table class="ExampleTable">
				<tr>
					<th>Gene</th>
					<th>Level_0</th>
					<th>Level_1</th>
					<th>etc.</th>
				</tr>
				<tr>
					<td>Gene1_Name</td>
					<td>Metabolism</td>
					<td>Lipid Metabolism</td>
					<td>...</td>
				</tr>
				<tr>
					<td>Gene2_Name</td>
					<td>Cellular Processes</td>
					<td>Cell Motility</td>
					<td>...</td>
				</tr>
				<tr>
					<td>etc.</td>
					<td>...</td>
					<td>...</td>
					<td>...</td>
				</tr>
			</table>
		</td>
	</tr>
	<tr>
		<td>
			Sample Map
			<input type="file" name="sample_map_C" id="sample_map_C">
		</td>
		<td>
			A file describing the relationship between samples, i.e. which ones are the same treatment condition.
		</td>
		<td>
			<table class="ExampleTable">
				<tr>
					<th>Sample</th>
					<th>Grouping_1</th>
					<th>Grouping_2</th>
					<th>etc.</th>
				</tr>
				<tr>
					<td>Sample1_Name</td>
					<td>Case</td>
					<td>Human</td>
					<td>...</td>
				</tr>
				<tr>
					<td>Sample2_Name</td>
					<td>Control</td>
					<td>Mouse</td>
					<td>...</td>
				</tr>
				<tr>
					<td>etc.</td>
					<td>...</td>
					<td>...</td>
					<td>...</td>
				</tr>
			</table>
		</td>
	</tr>
	<tr>
		<td>
			Function Abundances
			<input type="file" name="function_abundances_C" id="function_abundances_C">
		</td>
		<td>
			A file containing measured abundances of each function in each sample for comparison to function abundances predicted from taxonomic abundances.
		</td>
		<td>
			<table class="ExampleTable">
				<tr>
					<th>Gene</th>
					<th>Sample1_Name</th>
					<th>Sample2_Name</th>
					<th>etc.</th>
				</tr>
				<tr>
					<td>Gene1_Name</td>
					<td>0.1</td>
					<td>0.2</td>
					<td>...</td>
				</tr>
				<tr>
					<td>Gene2_Name</td>
					<td>0.3</td>
					<td>0.25</td>
					<td>...</td>
				</tr>
				<tr>
					<td>etc.</td>
					<td>...</td>
					<td>...</td>
					<td>...</td>
				</tr>
			</table>
		</td>
	</tr>

	<tr>
		<td colspan=3>
			<b>Visualization Options</b>
		</td>
	</tr>
	<tr>
		<td>
			Taxonomic level of detail
			<br>
			<select name="taxlevelofdetail_C" id="taxLODselector_C"></select>
		</td>
		<td colspan=2>
			Select the lowest level of taxonomic detail to be displayed. Going to the OTU level will significantly slow down the program.
		</td>
	</tr>
	<tr>
		<td>
			Function level of detail
			<br>
			<select name="funclevelofdetail_C" id="funcLODselector_C"></select>
		</td>
		<td colspan=2>
			Select the lowest level of function detail to be displayed. Going to the KO level is not recommended as single KOs can be mapped to multiple higher levels of function detail in the default hierarchy.
		</td>
	</tr>
	<tr>
		<td>
			Sample grouping
			<br>
			<select name="sampgroup_C" id="sampgroupselector_C"></select>
		</td>
		<td colspan=2>
			Select a grouping for samples if desired. Requires uploading a sample map.
		</td>
	</tr>
	<tr>
		<td>
			Color scale
			<br>
			<select name="color_option_selector_C" id="color_option_selector_C">
			<option value="Categories">Hierarchical color scale</option>
			<option value="Random">Random color scale</option>
			</select>
		</td>
		<td colspan=2>
			Color scheme. Random color scale will help discern closely related taxa but hierarchical color scale will look more pleasing.
		</td>
	</tr>
	<tr>
		<td colspan=3>
			<h4>Draw the burrito</h4>
		</td>
	</tr>
	<tr>
		<td colspan=3>
			<button id="update_button_C" type="button" class="btn action-button btn-large btn-primary">Assemble the Burrito</button>
		</td>
	</tr>
</table>

</div>
<button class="inputSelectorB" id="GenomeOption">
	<b>Genome Annotations</b>
	<br/><br/>
	Choose this option if you have an OTU table from 16S reads, and genome content for each OTU.
</button>

<div id="GenomeMenu" class="hidexp">
<table class="uploadtable" width=100%>
	<col>
	<col>
	<col align="right">
	<tr>
		<td colspan=3>
			<b>File Upload</b>
		</td>
	</tr>
	<tr>
		<td>
			<b>Genome Abundances<b>
			<br>
			<input type="file" name="taxonomic_abundances_1" id="taxonomic_abundances_1">
		</td>
		<td>
			A file containing abundances of each taxon in each sample. See documentation for an example.
		</td>
		<td>
			<table class="ExampleTable">
				<tr>
					<th>OTU</th>
					<th>Sample1_Name</th>
					<th>Sample2_Name</th>
					<th>etc.</th>
				</tr>
				<tr>
					<td>OTU1_ID</td>
					<td>10</td>
					<td>5</td>
					<td>...</td>
				</tr>
				<tr>
					<td>OTU2_ID</td>
					<td>0</td>
					<td>5</td>
					<td>...</td>
				</tr>
				<tr>
					<td>etc.</td>
					<td>...</td>
					<td>...</td>
					<td>...</td>
				</tr>
			</table>
		</td>
	</tr>
	<tr>
		<td>
			<div id="read_counts_g">
				<b>Genome Annotations<b>
				<br>
				<input type="file" name="genome_annotations" id="genome_annotations">
			</div>
		</td>
		<td>
			A file containing abundances of each taxon in each sample. See documentation for an example.
		</td>
		<td>
			<table class="ExampleTable">
				<tr>
					<th>OTU</th>
					<th>Gene</th>
					<th>CopyNumber</th>
				</tr>
				<tr>
					<td>OTU1_ID</td>
					<td>Gene1_Name</td>
					<td>1</td>
				</tr>
				<tr>
					<td>OTU1_ID</td>
					<td>Gene2_Name</td>
					<td>2</td>
				</tr>
				<tr>
					<td>etc.</td>
					<td>etc.</td>
					<td>etc.</td>
				</tr>
			</table>
		</td>
	</tr>
	<tr>
		<td>
			Taxonomic Hierarchy
			<input type="file" name="taxonomic_hierarchy_G" id="taxonomic_hierarchy_G">
		</td>
		<td>
			A file containing all taxa in your abundance file, in a hierarchy showing their relatedness. See documentation for a format example. Uses greengenes by default if no file uploaded.
		</td>
		<td>
			<table class="ExampleTable">
				<tr>
					<th>OTU</th>
					<th>Taxa_Level_1</th>
					<th>Taxa_Level_2</th>
					<th>etc.</th>
				</tr>
				<tr>
					<td>OTU1_ID</td>
					<td>Bacteria</td>
					<td>Firmicutes</td>
					<td>...</td>
				</tr>
				<tr>
					<td>OTU2_ID</td>
					<td>Bacteria</td>
					<td>Proteobacteria</td>
					<td>...</td>
				</tr>
				<tr>
					<td>etc.</td>
					<td>...</td>
					<td>...</td>
					<td>...</td>
				</tr>
			</table>
		</td>
	</tr>
	<tr>
		<td>
			Function Hierarchy
			<input type="file" name="function_hierarchy_G" id="function_hierarchy_G">
		</td>
		<td>
			A file containing the functions in your contribution or genome content file, in a hierarchy showing their relatedness. See documentation for a format example. Uses BRITE by default if no file is uploaded.
		</td>
		<td>
			<table class="ExampleTable">
				<tr>
					<th>Gene</th>
					<th>Level_0</th>
					<th>Level_1</th>
					<th>etc.</th>
				</tr>
				<tr>
					<td>Gene1_Name</td>
					<td>Metabolism</td>
					<td>Lipid Metabolism</td>
					<td>...</td>
				</tr>
				<tr>
					<td>Gene2_Name</td>
					<td>Cellular Processes</td>
					<td>Cell Motility</td>
					<td>...</td>
				</tr>
				<tr>
					<td>etc.</td>
					<td>...</td>
					<td>...</td>
					<td>...</td>
				</tr>
			</table>
		</td>
	</tr>
	<tr>
		<td>
			Sample Map
			<input type="file" name="sample_map_G" id="sample_map_G">
		</td>
		<td>
			A file describing the relationship between samples, i.e. which ones are the same treatment condition.
		</td>
		<td>
			<table class="ExampleTable">
				<tr>
					<th>Sample</th>
					<th>Grouping_1</th>
					<th>Grouping_2</th>
					<th>etc.</th>
				</tr>
				<tr>
					<td>Sample1_Name</td>
					<td>Case</td>
					<td>Human</td>
					<td>...</td>
				</tr>
				<tr>
					<td>Sample2_Name</td>
					<td>Control</td>
					<td>Mouse</td>
					<td>...</td>
				</tr>
				<tr>
					<td>etc.</td>
					<td>...</td>
					<td>...</td>
					<td>...</td>
				</tr>
			</table>
		</td>
	</tr>
	<tr>
		<td>
			Function Abundances
			<input type="file" name="function_abundances_G" id="function_abundances_G">
		</td>
		<td>
			A file containing measured abundances of each function in each sample for comparison to function abundances predicted from taxonomic abundances.
		</td>
		<td>
			<table class="ExampleTable">
				<tr>
					<th>Gene</th>
					<th>Sample1_Name</th>
					<th>Sample2_Name</th>
					<th>etc.</th>
				</tr>
				<tr>
					<td>Gene1_Name</td>
					<td>0.1</td>
					<td>0.2</td>
					<td>...</td>
				</tr>
				<tr>
					<td>Gene2_Name</td>
					<td>0.3</td>
					<td>0.25</td>
					<td>...</td>
				</tr>
				<tr>
					<td>etc.</td>
					<td>...</td>
					<td>...</td>
					<td>...</td>
				</tr>
			</table>
		</td>
	</tr>

	<tr>
		<td colspan=3>
			<b>Visualization Options</b>
		</td>
	</tr>
	<tr>
		<td>
			Taxonomic level of detail
			<br>
			<select name="taxlevelofdetail_G" id="taxLODselector_G"></select>
		</td>
		<td colspan=2>
			Select the lowest level of taxonomic detail to be displayed. Going to the OTU level will significantly slow down the program.
		</td>
	</tr>
	<tr>
		<td>
			Function level of detail
			<br>
			<select name="funclevelofdetail_G" id="funcLODselector_G"></select>
		</td>
		<td colspan=2>
			Select the lowest level of function detail to be displayed. Going to the KO level is not recommended as single KOs can be mapped to multiple higher levels of function detail in the default hierarchy.
		</td>
	</tr>
	<tr>
		<td>
			Sample grouping
			<br>
			<select name="sampgroup_G" id="sampgroupselector_G"></select>
		</td>
		<td colspan=2>
			Select a grouping for samples if desired. Requires uploading a sample map.
		</td>
	</tr>
	<tr>
		<td>
			Color scale
			<br>
			<select name="color_option_selector_G" id="color_option_selector_G">
			<option value="Categories">Hierarchical color scale</option>
			<option value="Random">Random color scale</option>
			</select>
		</td>
		<td colspan=2>
			Color scheme. Random color scale will help discern closely related taxa but hierarchical color scale will look more pleasing.
		</td>
	</tr>
	<tr>
		<td colspan=3>
			<h4>Draw the burrito</h4>
		</td>
	</tr>
	<tr>
		<td colspan=3>
			<button id="update_button_G" type="button" class="btn action-button btn-large btn-primary">Assemble the Burrito</button>
		</td>
	</tr>
</table>

</div>

<br>
<span class="error" id="upload_message"></span>
<button id="update_button" type="button" style="display:none" class="btn action-button btn-large btn-primary"></button>
</div>
<div style="float:left" id="button_holder"><br><br><br><br><br><br><br><br></div>

<div id="mainplot">

<svg id="patternsvg" height="0" width="0" xmlns="http://www.w3.org/2000/svg" version="1.1">
  <defs>
    <pattern id="crosshatch" patternUnits="userSpaceOnUse" width="4" height="4">
      <image xlink:href="data:image/svg+xml;base64,PHN2ZyB4bWxucz0naHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmcnIHdpZHRoPSc4JyBoZWlnaHQ9JzgnPgogIDxyZWN0IHdpZHRoPSc4JyBoZWlnaHQ9JzgnIGZpbGw9JyNmZmYnLz4KICA8cGF0aCBkPSdNMCAwTDggOFpNOCAwTDAgOFonIHN0cm9rZS13aWR0aD0nMC41JyBzdHJva2U9JyNhYWEnLz4KPC9zdmc+Cg==" x="0" y="0" width="4" height="4">
      </image>
    </pattern>
	
	<pattern id="rarrowspat" patternUnits="userSpaceOnUse" width="30" height = "20">
	    <rect x="0" y="0" width="20" height="20" fill="none" />
		<path d="M 0 10 L 17 10 L 10 17 L 17 10 L 10 3" fill="none" stroke="#D7764A" stroke-width="4" />
	</pattern>
	<pattern id="larrowspat" patternUnits="userSpaceOnUse" width="30" height = "20">
	    <rect x="0" y="0" width="30" height="30" fill="none" />
		<path d="M 20 10 L 3 10 L 10 17 L 3 10 L 10 3" fill="none" stroke="#D7764A" stroke-width="4" />
	</pattern>
  </defs>
</svg>

</div>

<script src="Javascript/constants.js"></script>
<script src="Javascript/textures.js"></script>
<script src="Javascript/d3.v3.min.js"></script>
<script src="Javascript/data_cube.js"></script>
<script src="Javascript/bp2.js"></script>
<script src="Javascript/FuncBar.js"></script>
<script src="Javascript/otu_stacked_bar.js"></script>
<script src="Javascript/colorbrewer.js"></script>
<script src="Javascript/upload.js"></script>
<script src="Javascript/mainui.js"></script>
<script src="Javascript/saveSvgAsPng.js"></script>
<script src="Javascript/trees.js"></script>
<script src="Javascript/helpOverlay.js"></script>
<script src="Javascript/main.js"></script>


</body>
