<!DOCTYPE html>
<head>
<meta charset="utf-8">

<title>Burrito Visualization</title>

<script src="shared/jquery.js" type="text/javascript"></script>
<script src="shared/shiny.js" type="text/javascript"></script>
<link rel="stylesheet" type="text/css" href="shared/shiny.css"/>
<script src="__assets__/shiny-server.js"></script>
<link rel="stylesheet" type="text/css" href="CSS/burritostyle.css"/>

</head>

<body>
<div id="mainthing" style="overflow-x:auto">


<table width="900" align="center">
	<tr>
		<td colspan=2 align="center">
			<img border="0" src="images/title_lab_800.jpg"></img>
		</td>
	</tr>
	<tr>
		<td colspan=2 align="center"> 
			<h2>Burrito: A Marriage of Taxonomy and Function</h2> 
			<img border="0" src="images/burrito.jpg"></img>
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> A tasty burrito.</span></span></span>
			
		</td>
	</tr>
	
	<tr>
		<td colspan=2>
			<p>This is a tool to visualize the relationship between taxonomy and function. You can view sample data or upload your own data.</p>
			<p>Source code available <a href="https://github.com/borenstein-lab/burrito">here</a></p>
		</td>
	</tr>
	
	<tr>
		<td>
			<table id="sampletable" width="400" height="200" frame="box">
			<tr><td>
				<input type="radio" name="whichinput" id="usesample" onchange="mainui.checkWhichInput()">
				<h4 style="display:inline">Explore sample data</h4> 
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> Select this option to use the tool to explore sample data.</span></span></span>
			</td></tr>
			<tr><td>
				<select name="sampleset" id="sampleselector">
					<option value="none" selected></option>
					<option value="theriot">Theriot et. al. 2014</option>
				</select>
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> Choose which data set to use. After selecting one a description of it will be shown below.</span></span></span>
				
			</td></tr>
			<tr><td id="samplesetinfo">
				<div id="noneexp">
					Information about the selected dataset.
				</div>
				<div id="theriotexp" class="hidexp">
					<p>Antibiotic-induced shifts in the mouse gut microbiome and metabolome increase susceptibility to Clostridium difficile infection</p>
					<a href=http://www.nature.com/ncomms/2014/140120/ncomms4114/abs/ncomms4114.html target="_blank">Paper</a>
				</div>
			</td></tr>
			</table>
		</td>
		<td>
			<table id="uploadtable" width="500" height = "200" frame="box">
				<tr><td colspan=3>
					<input type="radio" name="whichinput" id="useupload" onchange="mainui.checkWhichInput()">
					<h4 style="display:inline">Upload your own data</h4>
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> Select this option if you have your own data to upload and visualize.</span></span></span>
					 
				</td></tr>
				<tr><td colspan=3>
						<input type="radio" name="inptype" id="sixteenradio" onchange="mainui.checkInpDataRadio()" value="16s">16s Reads
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> You only have 16s reads.</span></span></span>
						
						<input type="radio" name="inptype" id="funcradio" onchange="mainui.checkInpDataRadio()" value="func">Functional Contributions
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> You have run PICRUST yourself and have reads and contributions.</span></span></span>
						
						<input type="radio" name="inptype" id="genomeradio" onchange="mainui.checkInpDataRadio()" value="genome">Genomic Annotations
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> You have the genome content of each taxa and their abundances.</span></span></span>
				
				</td></tr>
				<tr>
					<td>
						<div id="taxonomic_abundances_1_g" class="hidbut">
							Genome Abundances
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> A file containing abundances of each genome in each sample. See documentation for an example.</span></span></span>
							<input type="file" name="taxonomic_abundances_1" id="taxonomic_abundances_1">
						</div>
						<div id="taxonomic_abundances_2_g" class="hidexp">
							Taxonomic Abundances
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> A file containing abundances of each taxa in each sample. See documentation for an example.</span></span></span>
							<input type="file" name="taxonomic_abundances_2" id="taxonomic_abundances_2">
						</div>
						<div id="read_counts_g" class="hidexp">
							16s Reads
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> A file containing abundaces of each taxa in each sample. See documentation for an example.</span></span></span>
							<input type="file" name="read_counts" id="read_counts">
						</div>
					</td>
					<td>
						<div id="genome_annotations_g" class="hidbut">
							Genome Annotations
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> A file containing the functional content of each genome. See documentation for an example.</span></span></span>
							<input type="file" name="genome_annotations" id="genome_annotations">
						</div>
						<div id="function_contributions_g" class="hidexp">
							Function Contributions
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> A file containing the functional contribution of each taxa. See documentation for an example.</span></span></span>
							<input type="file" name="function_contributions" id="function_contributions">
						</div>
					</td>
					<td>
						
					</td>
				</tr>
				<tr> 
					<td colspan=3> Optional Files:</td>
				</tr>
				<tr>
					<td>
						Taxonomic Hierarchy
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> A file containing all taxa in your abundance file, in a hierarchy showing their relatedness. See documentation for a format example. Uses greengenes by default if no file uploaded.</span></span></span>
						<input type="file" name="taxonomic_hierarchy" id="taxonomic_hierarchy">
					</td>
					<td>
						Function Hierarchy
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> A file containing the functions in your contribution or genome content file, in a hierarchy showing their relatedness. See documentation for a format example. Uses BRITE by default if no file is uploaded.</span></span></span>
						<input type="file" name="function_hierarchy" id="function_hierarchy">
					</td>
					<td>
						Sample Map
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> A file describing the relationship between samples, i.e. which ones are the same treatment condition.</span></span></span>
						<input type="file" name="sample_map" id="sample_map">
					</td>
				</tr>
			</table>
		</td>
	</tr>
	<tr> <td width="15" bgcolor="000000" colspan=2> 
	</td></tr>
	<tr>
		<td colspan=2> <h4>Visualization Options</h4> 
			<select name="taxlevelofdetail" id="taxLODselector">
			</select>
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> Select the lowest level of taxonomic detail to be displayed. Going to the OTU level will significantly slow down the program.</span></span></span>
			
		</td>
		</tr>
		<tr>
		<td>
			<select name="color_option_selector" id="color_option_selector">
			<option value="Categories">Hierarchical color scale</option>
			<option value="Random">Random color scale</option>
			</select>
				<span class="tt"><img src="images/help.png" style="vertical-align; middle;"><span class="tooltip"><span class="blank"></span><span class="content"> Random color scale will help discern closely related taxa but hierarchical color scale will look more pleasing.</span></span></span>
			
		</td>
	</tr>

	<tr>
		<td colspan=2>
			<h4>Draw the burrito</h4>
			<button id="update_button" type="button" class="btn action-button btn-large btn-primary">Generate Visualization</button>
			<button id="update_button5000" type="button" class="btn action-button btn-large btn-primary" style="display:none"></button>
		</td>
	</tr>
</table>


<br>
<span class="error" id="upload_message"></span>
</div>
<div style="float:left" id="button_holder"><br><br><br><br><br><br><br><br></div>

<div id="mainplot">



<svg id="patternsvg" height="0" width="0" xmlns="http://www.w3.org/2000/svg" version="1.1">
  <defs>
    <pattern id="crosshatch" patternUnits="userSpaceOnUse" width="4" height="4">
      <image xlink:href="data:image/svg+xml;base64,PHN2ZyB4bWxucz0naHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmcnIHdpZHRoPSc4JyBoZWlnaHQ9JzgnPgogIDxyZWN0IHdpZHRoPSc4JyBoZWlnaHQ9JzgnIGZpbGw9JyNmZmYnLz4KICA8cGF0aCBkPSdNMCAwTDggOFpNOCAwTDAgOFonIHN0cm9rZS13aWR0aD0nMC41JyBzdHJva2U9JyNhYWEnLz4KPC9zdmc+Cg==" x="0" y="0" width="4" height="4">
      </image>
    </pattern>
	
	<pattern id="rarrowspat" patternUnits="userSpaceOnUse" width="30" height = "20">
	    <rect x="0" y="0" width="20" height="20" fill="none" />
		<path d="M 0 10 L 17 10 L 10 17 L 17 10 L 10 3" fill="none" stroke="#D7764A" stroke-width="4" />
	</pattern>
	<pattern id="larrowspat" patternUnits="userSpaceOnUse" width="30" height = "20">
	    <rect x="0" y="0" width="30" height="30" fill="none" />
		<path d="M 20 10 L 3 10 L 10 17 L 3 10 L 10 3" fill="none" stroke="#D7764A" stroke-width="4" />
	</pattern>
  </defs>
</svg>

</div>

<script src="Javascript/textures.js"></script>
<script src="Javascript/d3.v3.min.js"></script>
<script src="Javascript/data_cube.js"></script>
<script src="Javascript/bp2.js"></script>
<script src="Javascript/FuncBar.js"></script>
<script src="Javascript/otu_stacked_bar.js"></script>
<script src="Javascript/colorbrewer.js"></script>
<script src="Javascript/upload.js"></script>
<script src="Javascript/buttons.js"></script>
<script src="Javascript/input_parsing.js"></script>
<script src="Javascript/mainui.js"></script>
<script src="Javascript/saveSvgAsPng.js"></script>
<script src="Javascript/trees.js"></script>

<script>
mainui.createUI();

var aspecrat, width, height, hidewidth, margin;
var barDimensions, navDims, bpdims;
var i,
duration;

//var diagonal;

var MainSVG, plotSVG;

draw_svg = function() {
	if (d3.select("#mainsvg")[0][0] === null) {
		aspecrat = window.innerWidth / window.innerHeight;

		width = 1000 * aspecrat;
		height = 1000;
		hidewidth = 0;
		margin ={left:(10 + hidewidth), right:20, top:20, bottom:20 , btwnavbar:40, btwbars:40};
		barDimensions = {width: (width/2 - (margin.btwbars/2) - margin.left), height: (height / 2) - margin.bottom - (margin.btwnavbar/2) };
		navDims = {width: (width - margin.left - margin.right), height: (height/2) - margin.top - (margin.btwbars/2)};
		navDims.treewidth = navDims.width * 2/9;
		bpdims = {height:navDims.height, width: navDims.width, header:margin.top};
		i = 0;
		duration = 300;
		roots = {};

		trees.SetUp(navDims);
		

/*		diagonal = d3.svg.diagonal()  DELETE IF BEING COMMENTED ISNT A PROBLEM
			.projection(function (d) {
			return [d.y, d.x];
		}); */

		MainSVG = d3.select("#mainplot").append("svg")
			.attr("id","mainsvg")
			.attr("viewBox","0 0 " + (width - hidewidth) + " " + height + "")
			.attr("preserveAspectRatio","none");

		uploader.svgCreated = true;	
		plotSVG = d3.select("#mainsvg").append("svg")
			.attr("id","plots_svg")
			.attr("viewBox","0 0 " + (width - hidewidth) + " " + height + "")
			.attr("preserveAspectRatio","none")
			.style("font-family", "Verdana");
	
		button_maker.add_rect_button(d3.select("#mainsvg"), "save_png", 10, 10, 80, 80, "switchb activebutton", fill = "lightgrey", fontsize = "16px", "Save screenshot as png", function(){
			saveSvgAsPng(document.getElementById("plots_svg"), "screenshot.png", { backgroundColor : "white", scale: 1.5})
		});
		button_maker.add_rect_button(d3.select("#mainsvg"), "save_svg", 10, 100, 80, 80, "switchb activebutton", fill = "lightgrey", fontsize = "16px", "Save screenshot as svg", function(){
			svgAsDataUri(document.getElementById("plots_svg"), {}, function(uri) { //console.log(uri)
				saveSvg(uri, "screenshot.svg")
			}); 
		});
		button_maker.add_rect_button(d3.select("#mainsvg"), "save_png_taxa_bar", 10, 190, 80, 80, "switchb activebutton", fill = "lightblue", fontsize = "14px", "Save taxonomy plot as png", function(){
			saveSvgAsPng(document.getElementById("taxa_bars"), "taxa_bar.png",{ backgroundColor : "white", scale: 1.5})
		});
		button_maker.add_rect_button(d3.select("#mainsvg"), "save_svg_taxa_bar", 10, 280, 80, 80, "switchb activebutton", fill = "lightblue", fontsize = "14px", "Save taxonomy plot as svg", function(){
			svgAsDataUri(document.getElementById("taxa_bars"), {}, function(uri) {
				saveSvg(uri, "taxa_bar.svg")
			}); 
		});
		button_maker.add_rect_button(d3.select("#mainsvg"), "save_taxonomy_leg", 10, 370, 80, 50, "switchb activebutton", fill = "lightblue", fontsize = "14px", "Save taxonomy legend", function(){
			saveSvgAsPng(document.getElementById("saveLegBar0"), "taxonomy_legend.png",{ backgroundColor : "white", scale: 1.5})
		});
		button_maker.add_rect_button(d3.select("#mainsvg"), "save_taxonomy_leg_svg", 10, 430, 80, 50, "switchb activebutton", fill = "lightblue", fontsize = "14px", "Save taxonomy legend svg", function(){
			svgAsDataUri(document.getElementById("saveLegBar0"), {}, function(uri) {
				saveSvg(uri, "taxonomy_legend.svg")
			}); 
		});
		button_maker.add_rect_button(d3.select("#mainsvg"), "save_png_func_bar", 10, 490, 80, 80, "switchb activebutton", fill = "lightgreen", fontsize = "14px", "Save function plot as png", function(){
			saveSvgAsPng(document.getElementById("func_bars"), "func_bar.png",{ backgroundColor : "white", scale: 1.5})
		});
		button_maker.add_rect_button(d3.select("#mainsvg"), "save_svg_func_bar", 10, 580, 80, 80, "switchb activebutton", fill = "lightgreen", fontsize = "14px", "Save function plot as svg", function(){
			svgAsDataUri(document.getElementById("func_bars"), {}, function(uri) {
				saveSvg(uri, "function_bar.svg")
			});
		});
		button_maker.add_rect_button(d3.select("#mainsvg"), "save_function_leg", 10, 670, 80, 50, "switchb activebutton", fill = "lightgreen", fontsize = "14px", "Save function legend", function(){
			saveSvgAsPng(document.getElementById("saveLegBar1"), "function_legend.png",{ backgroundColor : "white", scale: 1.5}) });
		button_maker.add_rect_button(d3.select("#mainsvg"), "save_function_leg_svg", 10, 730, 80, 50, "switchb activebutton", fill = "lightgreen", fontsize = "14px", "Save function legend svg", function(){
			svgAsDataUri(document.getElementById("saveLegBar1"), {}, function(uri) {
				saveSvg(uri, "function_legend.svg")
			});
		});
		draw_loading();

	}
}

draw_loading = function() {
	var loadg = MainSVG.append("g")
		.attr("id","loadingG");
	
	loadg.append("text")
		.attr("x",width / 2)
		.attr("y", height / 2)
		.attr("text-anchor","middle")
		.text("Loading");

	var loadspin = loadg.append("ellipse")
		.attr("id","loadcirc")
		.attr("cx",(width / 2) - 100)
		.attr("cy",height / 2)
		.attr("rx", 30)
		.attr("ry", 30)
		.attr("fill","red")
		.attr("stroke","none");
}

draw_everything = function(otu_table, contribution_table, tax_hierarchy_text, func_hierarchy_text, samp_map_text, func_averages){
	
	// Find the new window size, adjust the aspect ratio
	aspecrat = window.innerWidth / window.innerHeight;
	width = 1000 * aspecrat;
	barDimensions = {width: (width/2 - (margin.btwbars/2) - margin.left), height: (height / 2) - margin.bottom - (margin.btwnavbar/2) };
	navDims = {width: (width - margin.left - margin.right), height: (height/2) - margin.top - (margin.btwbars/2)};
	navDims.treewidth = navDims.width * 2/9;
	bpdims = {height:navDims.height, width: navDims.width, header:margin.top};
	
	MainSVG.attr("viewBox","0 0 " + width + " " + height + "")
	plotSVG.attr("viewBox","0 0 " + width + " " + height + "")
	
	d3.select("#navbar").remove()
	d3.select("#taxa_bars").remove()
	d3.select("#func_bars").remove()
	d3.select("#loadingG").remove()	
	d3.select("#navbar").remove()
	d3.select("#taxa_bars").remove()
	d3.select("#func_bars").remove()

	var NavSVG = plotSVG.insert("svg", "#sidebar")
    	.attr("x",margin.left)
	.attr("y",margin.top)
	.attr("width", width - margin.left - margin.right)
	.attr("height", (height/2) - margin.top - (margin.btwnavbar / 2))
	.attr("id","navbar")
	.style("font-family", "Verdana");
			
	var bpG = NavSVG.append("g")
		.attr("transform","translate("+ (width / 2) +","+ margin.top +")");
		
	trees.SetUp2("navbar", margin, navDims, tax_hierarchy_text, func_hierarchy_text);
	
	var TaxaBar = plotSVG.insert("svg", "#sidebar")
		.attr("x",margin.left)
		.attr("y",(height/2) + (margin.btwnavbar/2))
		.attr("width",(width/2) - margin.left - (margin.btwbars/2))
		.attr("height",(height/2) - margin.bottom - (margin.btwnavbar/2))
		//.attr("width",100)
		//.attr("height",100)
		//.attr("viewBox","0 0 100 100")
		.attr("id", "taxa_bars")
		.style("font-family", "Verdana");
		//.attr("preserveAspectRatio","none");
		//.attr("transform","scale(2.0)");
		
	var FunctionBar = plotSVG.insert("svg", "#sidebar")
		.attr("x",(width/2) + (margin.btwbars/2))
		.attr("y",(height/2) + (margin.btwnavbar/2))
		.attr("width",(width/2) - margin.right - (margin.btwbars/2))
		.attr("height",(height/2) - margin.bottom - (margin.btwnavbar/2))
		//.attr("viewBox","0 0 100 100")
		.attr("id", "func_bars")
		.style("font-family", "Verdana");
		
	var bpvisdata;
	
	
	func_averages = d3.tsv.parse(func_averages);
	samplemap = d3.tsv.parse(samp_map_text);

	/*	otu_abundance_data_transpose = d3.tsv.parse(tax_abund_text);
	var fix_otus = function(otu_data, samples){
		new_otu_data = []
		for(j=0; j < samples.length; j++){
			new_otu_data[j] = {}
			new_otu_data[j]["Sample"] = samples[j]
			for(k=0; k < otu_data.length; k++){
				new_key = "OTU_ID_"+otu_data[k]["OTU_ID"] // Can we key anonymously? Maybe set name of column ourselves so it always matches
				new_otu_data[j][new_key] = +otu_data[k][samples[j]]
			}

		}	
		return new_otu_data;
	}

	otu_abundance_data = fix_otus(otu_abundance_data_transpose, d3.keys(contribution_table).sort()) //this puts samples in alphabetical order currently*/
	
	otu_abundance_data = otu_table

	//initialize display cube
	var data_cube = data_cube_wrapper.make_cube(); //defines the functions needed

	data_cube.initialize_cube(contribution_table, trees.getTaxaTreeData(), trees.getFuncTreeData(), func_averages);
			
	//if( not expand to OTU level
	/*genus_abundance_data = data_cube.reduce_to_genus(otu_abundance_data)*/

	////////////////////////// Colors
	//color_option = "Categories" //Categories or Random
	color_option = d3.select("#color_option_selector").property("value")
	console.log(color_option)
	var setUpColorScale = function(main_list, list_type, color_scale){ //list_type is taxa or funcs
		//1st - divide up color space into number of leaves
		//then assign each subgroup of leaves alternating extremes
		//then assign each parent extreme or average value of its children
		for(j=0; j < main_list.length; j++){
			if(list_type=="taxa"){
				var leaves = data_cube.get_leaves(main_list[j], data_cube.taxa_lookup)
			} else {
				var leaves = data_cube.get_leaves(main_list[j], data_cube.func_lookup)
			}
		    color_range = leaves.length //number of colors to expand in each direction depends on # of leaves

		    var core_color = d3.hcl(color_scale(main_list[j]))
	    	new_colors = []
	    	flip = -1
	    	for(k=0; k < color_range; k++){
	    		var new_color = d3.hcl()
				new_color["l"] = core_color["l"] - 36*flip*(k + (k%2))/(color_range+1)
	      		new_color["h"] = core_color["h"] - 8*flip*(k+ (k%2))/(color_range+1)
	      		new_color["c"] = core_color["c"] - 4*flip*(k+ (k%2))/(color_range+1)
	      		new_colors.push(new_color)
	      		flip *= -1

	      	}
			new_colors = new_colors.sort(function(a,b){ return a.l - b.l })

	      	if(list_type=="taxa"){
	      		layer1 = data_cube.taxa_lookup[main_list[j]]
	      		descendents = data_cube.get_descendents(main_list[j], data_cube.taxa_lookup)
	      		first_level = descendents.filter(function(d){ if(layer1.level != 1){ return data_cube.taxa_lookup[d].level == 1 } else { return data_cube.taxa_lookup[d].level == 0 }})
	      	} else {
	      		layer1 = data_cube.func_lookup[main_list[j]]
	      		descendents = data_cube.get_descendents(main_list[j], data_cube.func_lookup)
	      		first_level = descendents.filter(function(d){ return data_cube.func_lookup[d].level == 1})
	      	}
	      	num_levels = layer1.level

	      	color_count = 0	      	
	      	for(k=0; k < first_level.length; k++){
	      		if(list_type=="taxa"){
	      			desc_leaves = data_cube.get_leaves(first_level[k], data_cube.taxa_lookup)
	      		} else {
	      			desc_leaves = data_cube.get_leaves(first_level[k], data_cube.func_lookup)	      			
	      		}
	      		color_slice = new_colors.slice(color_count, color_count + desc_leaves.length) 
	      		color_count += desc_leaves.length
	      		//assign variants to each group of leaves
	      		for(m=0; m < desc_leaves.length; m++){
	      			if(color_scale.domain().indexOf(desc_leaves[m])==-1){
	      				if(m % 2==0){ //go from front or back
			    			color_scale.range().push(color_slice[m])
	      				} else {
	      					color_scale.range().push(color_slice[desc_leaves.length-m])
	      				}
	      			}
	      		}
	      	}
	      	//now get the rest of the levels, assign mean or most extreme value of descendents
	      	level = 1
	      	while(level < num_levels){
	      		if(list_type=="taxa"){
	      			level_items = descendents.filter(function(d){ return data_cube.taxa_lookup[d].level == level })
	      		} else {
	      			level_items = descendents.filter(function(d){ return data_cube.func_lookup[d].level == level })
	      		}
	      		parent_color = core_color
	      		for(m=0; m < level_items.length; m++){
	      			if(list_type=="taxa"){
	      				desc_leaves = data_cube.get_leaves(level_items[m], data_cube.taxa_lookup)
	      			} else {
	      				desc_leaves = data_cube.get_leaves(level_items[m], data_cube.func_lookup)
	      			}
	      			desc_colors = desc_leaves.map(function(d){ return d3.hcl(color_scale(d)) })
	      		if(m==0 & level_items.length > 1){
	      			parent_color.h = d3.min(desc_colors.map(function(d){ return d.h}))
	      			parent_color.c = d3.min(desc_colors.map(function(d){ return d.c}))
	      			parent_color.l = d3.min(desc_colors.map(function(d){ return d.l}))
	      		} else if(m == level_items.length -1 & m != 0) {
	      			parent_color.h = d3.max(desc_colors.map(function(d){ return d.h}))
	      			parent_color.c = d3.max(desc_colors.map(function(d){ return d.c}))
	      			parent_color.l = d3.max(desc_colors.map(function(d){ return d.l}))
	      		} else{
					parent_color.h = d3.mean(desc_colors.map(function(d){ return d.h}))
	      			parent_color.c = d3.mean(desc_colors.map(function(d){ return d.c}))
	      			parent_color.l = d3.mean(desc_colors.map(function(d){ return d.l}))
	      		
	      		}
		      	if(color_scale.domain().indexOf(level_items[m])==-1){
		      		color_scale.range().push(parent_color.toString())
	      			color_scale(level_items[m])
	      		}
	      		}
	      		level++;
	      	}
	      }
		return color_scale;
	}

	if(color_option == "Categories"){
	
		//set up colors using data cube
		//colors	
		//start one level in for bacteria but at the 0th level for functions
		num_function_categories = data_cube.func_tree.length
		num_taxa_categories = 0
		for(j=0; j < data_cube.taxa_tree.length; j++){
			num_taxa_categories += data_cube.taxa_tree[j].values.length
		}

		//taxa colors
		if(num_taxa_categories > d3.keys(colorbrewer["Set3"]).pop()+d3.keys(colorbrewer["Dark2"]).pop()){
			console.log("too many taxa categories, colors will repeat")
		}
		if(num_taxa_categories <= d3.keys(colorbrewer["Set3"]).pop()){
			taxa_palette = colorbrewer["Set3"][num_taxa_categories]
			//taxa_palette = taxa_palette.reverse()
		} else {
			taxa_palette = colorbrewer["Set3"][(d3.keys(colorbrewer["Set3"]).pop()-1)].concat(colorbrewer["Dark2"][num_taxa_categories -  d3.keys(colorbrewer["Set3"]).pop()])
		}
		if(num_function_categories > d3.keys(colorbrewer["Set1"]).pop()+d3.keys(colorbrewer["Dark2"]).pop()){
			console.log("too many function categories, colors will repeat")
		}

		if(num_function_categories <= d3.keys(colorbrewer["Set1"]).pop()){
			func_palette = colorbrewer["Set1"][num_function_categories]
		} else {
			func_palette = colorbrewer["Set1"][(d3.keys(colorbrewer["Set1"]).pop()-1)].concat(colorbrewer["Dark2"][num_function_categories - d3.keys(colorbrewer["Set1"]).pop()])
		}

		var taxa_colors = d3.scale.ordinal()
		taxa_colors.range(taxa_palette)
		main_taxa = []
		for(j=0; j < data_cube.taxa_tree.length; j++){
			main_taxa = main_taxa.concat(data_cube.taxa_tree[j].values.map(function(d){ return d.key;}))
		}
		taxa_colors.domain(main_taxa)
		taxa_colors = setUpColorScale(main_taxa, "taxa", taxa_colors)

		kingdoms = data_cube.taxa_tree.map(function(d){ return d.key})
		col1 = d3.hsl("black")
		//if there are other kingdoms besides Bacteria, shades of grey
		for(j=0; j < kingdoms.length; j++){
			taxa_colors.range().push(col1)
			taxa_colors(kingdoms[j])
			col1["l"] +=  (j+1)/(kingdoms.length + 1)
		}
		taxa_colors.range().push("black")
		taxa_colors("All Taxa")

		//func colors
		var func_colors = d3.scale.ordinal()
		func_colors.range(func_palette)
		main_funcs = data_cube.func_tree.map(function(d){ return d.key;})
		func_colors.domain(main_funcs)
		func_colors = setUpColorScale(main_funcs, "funcs", func_colors)
		func_colors.range().push("black")
		func_colors("All Functions")
	
	} else if(color_option == "Random"){
		taxa_colors = d3.scale.category20()
		taxa_colors.domain(d3.keys(data_cube.taxa_tree_lookup))
		func_colors = d3.scale.ordinal()
		func_colors.range(colorbrewer["Set3"]["12"].concat(colorbrewer["Dark2"]["8"]))
		func_colors.domain(d3.keys(data_cube.func_tree_lookup))
	}


	//sample colors
	groupValsAll = samplemap.map(function(d,i){ 
		return d["Group"]; })

	groupVals = groupValsAll.filter(function(d,i){ return groupValsAll.indexOf(d)===i; })

	var sampleColor = d3.scale.ordinal().domain(groupVals).range(colorbrewer["Paired"][Math.max(4,groupVals.length)].slice(0,groupVals.length));
	
	getLinkData = function(){
		//get unique linked functions for each taxon
		allSamples = d3.keys(data_cube.displayed_contribution_cube);
		var all_taxa_links = []; //{Taxa:[], Functions:[]};
		data_cube.displayed_taxa.map(function(t,i){
			data_cube.displayed_funcs.map(function(f,i){
				var foo = allSamples.map(function(d){
					return data_cube.get_contribution(d,t,f);
				})
				if(d3.sum(foo)!==0){
					all_taxa_links.push({Taxa:t, Functions:f, numKO:1});
				}
			})
		});
		return all_taxa_links;
	}

	//this gets the data for the functional bar chart

	getFuncBarData = function(){
		var contributions = [];
		var samples = [];
		var funcs = [];
		var taxa = [];

		var all_func_data = [];
		var count = 0;
		var tcount = 0;

		allSamples = d3.keys(data_cube.displayed_contribution_cube).sort();
		var all_taxa_links = []; //{Taxa:[], Functions:[]};
		//loop through all the functions
		//this keeps all functions together

		//loop through Samples, funcs and taxa and create arrays of the data
		allSamples.map(function(d){	
			samples.push({sample:d});
		});
		var my_displayed_funcs = data_cube.displayed_funcs.slice(0);
		my_displayed_funcs.reverse();
		(my_displayed_funcs).map(function(z){
			funcs.push(z);
		});
		var my_displayed_taxa = data_cube.displayed_taxa.slice(0);
		my_displayed_taxa.reverse();
		(my_displayed_taxa).map(function(t,i){
			taxa.push(t)
		});
		//then loop through sequentially to populate an array of all the samples
		samples.map(function(n){
			funcs.map(function(m){

				taxa.map(function(tt){	
					contributions.push({func:m, Taxa:tt, Sample:n.sample, contributions:data_cube.get_contribution([n.sample],[tt],[m])});})
			})
		var temp = [];
		contributions.map(function(q){
			if (q.Sample == n.sample) {
				temp.push({func:q.func, Taxa:q.Taxa, Sample:q.Sample, contributions:q.contributions});}
		})

		all_func_data.push({data:temp, Sample:n.sample});
		count = count + 1
		});
		return all_func_data;
	}

	function sort_nest(x, y){
		if (x.key < y.key){
			return -1;
		} else if (x.key > y.key){
			return 1;
		} else {
			return 0;
		}
	}

	//data_cube.expand_taxon("Bacteria");

	var bpData = getLinkData();
	var data = {data:bP.partData(bpData, data_cube.displayed_taxa, data_cube.displayed_funcs), id:'Genomes', header:["Taxa","Functions", "Genomes"]};


	//draw the stacked bar

	var stackData = getFuncBarData();

	fB.Draw(stackData, samplemap, func_colors, FunctionBar, barDimensions, highlightOverall, dehighlightOverall, sampleColor);


	var otu_bar_data = otu_bar.make_data(otu_abundance_data, data_cube);

	otu_bar.draw(otu_bar_data, samplemap, taxa_colors, TaxaBar, barDimensions, highlightOverall, dehighlightOverall, sampleColor);


	bP.draw(data, bpG, bpdims, taxa_colors, func_colors, data_cube.displayed_taxa, data_cube.displayed_funcs,highlightOverall, dehighlightOverall);


	function update_otu_bar(){
		//remove old graph before redrawing new
		TaxaBar.selectAll("g").remove();
		var otu_bar_data = otu_bar.make_data(otu_abundance_data, data_cube);
		otu_bar.draw(otu_bar_data, samplemap, taxa_colors, TaxaBar, barDimensions, highlightOverall, dehighlightOverall, sampleColor);
	}

	function update_func_bar(){
		//remove old graph before redrawing new
			FunctionBar.selectAll("g").remove();
			var func_data = getFuncBarData();
			fB.Draw(func_data, samplemap, func_colors, FunctionBar, barDimensions, highlightOverall, dehighlightOverall, sampleColor);

	}

	//data_cube.expand_func("Metabolism");
	var bpData = getLinkData();
	var data = {data:bP.partData(bpData, data_cube.displayed_taxa, data_cube.displayed_funcs), id:'Genomes', header:["Taxa","Functions", "Genomes"]};
	bpvisdata = bP.updateGraph(data, bpG, bpdims, taxa_colors, func_colors, data_cube.displayed_taxa, data_cube.displayed_funcs, highlightOverall, dehighlightOverall);


    //d3.select(self.frameElement).style("height", "800px");

	trees.SetUp3(height, data_cube, otu_abundance_data, bpvisdata, highlightOverall, dehighlightOverall);



    var curr_funcs = [];
    for (var i = 0; i < function_data.length; i++){
	    curr_funcs.push(function_data[i]);
	}
    for (; curr_funcs.length > 0;){
    	curr_func = curr_funcs.shift();
    	if (!data_cube.is_leaf(curr_func)){
    		curr_func.values.sort(sort_nest);
    		for (var i = 0; i < curr_func.values.length; i++){
    			curr_funcs.push(curr_func.values[i]);
    		}
    	}
    }  

    

		
	function highlightOverall(taxonName, functionName, highlightwhat) {
		if (highlightwhat == 1) {
			// Highlight taxa tree
			trees.highlightTree(taxonName, 'taxa');
			
			// Highlight BP graph
			m = 0;
			displayed = data_cube.displayed_taxa;
			ind = displayed.indexOf(taxonName);	
			if (ind > -1) {
				bP.selectSegment(m, ind, taxa_colors, func_colors, data_cube.displayed_taxa, data_cube.displayed_funcs);
			
				// Highlight  Taxa Bar Chart
				otu_bar.select_bars(taxonName);

				// Highlight Function Bar Chart
				fB.select_contribution(taxonName, func_colors);
			}
		}
		
		if (highlightwhat == 2) {
			// Highlight function tree
			trees.highlightTree(functionName, 'func');
			
			// Highlight BP graph
			m = 1;
			displayed = data_cube.displayed_funcs;
			ind = displayed.indexOf(functionName);
			if (ind > -1) {
				bP.selectSegment(m, ind, taxa_colors, func_colors, data_cube.displayed_taxa, data_cube.displayed_funcs);

				// Highlight Function Bar Chart
				fB.select_bars(functionName, func_colors);
			}
		}
		
		if (highlightwhat == 3) {
			//highlight taxa tree
			trees.highlightTree(taxonName, 'taxa');
			
			//highlight function tree
			trees.highlightTree(functionName, 'func');
		
			// Highlight Taxa Bar Chart
			otu_bar.select_bars(taxonName);

			// Highlight Function Bar Chart
			fB.select_single_contribution(taxonName, functionName, func_colors);
		}
	}

	function dehighlightOverall(taxonName, functionName, highlightwhat) {

		if (highlightwhat == 1) {
			// dehighlight tree
			trees.dehighlightTree(taxonName, 'taxa');
			
			// dehighlight BP graph
			m = 0;
			displayed = data_cube.displayed_taxa;	
			ind = displayed.indexOf(taxonName);
			if (ind > -1) {
				bP.deSelectSegment(m, ind, taxa_colors, func_colors, data_cube.displayed_taxa, data_cube.displayed_funcs);

				// dehighlight Taxon Bar Chart
				otu_bar.deselect_bars(taxonName, taxa_colors);

				// dehighlight Function Bar Chart
				fB.deselect_contribution(taxonName, func_colors);
			}
		}
		
		if (highlightwhat == 2) {
			trees.dehighlightTree(functionName, 'func');
			
			//dehighlight BP graph
			m = 1;
			displayed = data_cube.displayed_funcs;
			ind = displayed.indexOf(functionName);
			if (ind > -1) {
				bP.deSelectSegment(m, ind, taxa_colors, func_colors, data_cube.displayed_taxa, data_cube.displayed_funcs);

				// dehighlight Function Bar Chart
				fB.deselect_bars(functionName, func_colors);
			}
		}
		
		if (highlightwhat == 3) {
			// dehighlight taxa tree
			trees.dehighlightTree(taxonName, 'taxa');
			
			//dehighlight function tree
			trees.dehighlightTree(functionName, 'func');
			
			// dehighlight Taxa Bar Chart
			otu_bar.deselect_bars(taxonName, taxa_colors);

			// dehighlight Function Bar Chart
			fB.deselect_single_contribution(taxonName, functionName, func_colors);
		
		}
	}
	
	// Update the bipartite graph after changes to data_cube
	function updateBPgraph() {
		var bpData = getLinkData();
		var data = {data:bP.partData(bpData, data_cube.displayed_taxa, data_cube.displayed_funcs), id:'Genomes', header:["Taxa","Functions", "Genomes"]};
		var visdata = bP.updateGraph(data, bpG, bpdims, taxa_colors, func_colors, data_cube.displayed_taxa, data_cube.displayed_funcs, highlightOverall, dehighlightOverall);
		return visdata;
	}
}

var uploader = uploader_wrapper.make_uploader(draw_everything);

mainui.uploadMode = "Function"; // Should not have to do this, try to eliminate later

Shiny.addCustomMessageHandler("shiny_test", function(message){console.log(message)})
	
</script>
</body>
