---
title: 'BURRITO: An Interactive Multi-Omic Tool for Visualizing Taxa-Function Relationships
  in Microbiome Data'
output: html_document
date: March 13, 2020
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

**URL:** http://elbo-spice.cs.tau.ac.il/shiny/burrito/

**Documentation:** http://borenstein-lab.github.io/burrito

**Publication:** McNally, C.P., Eng, A., Noecker, C., Gagne-Maynard, W.C., and Borenstein, E. (2018). BURRITO: An Interactive Multi-Omic Tool for Visualizing Taxa-Function Relationships in Microbiome Data. Front. Microbiol. 9, 365.

## Quick Start Summary

Follow these steps to visualize two example vaginal microbiome datasets with BURRITO.

1) Download the example data by clicking and unzipping [**this file**](https://cnoecker.github.io/TAUMicrobiomeWorkshop2019/data/BurritoData.zip). See below for more details on these datasets.

2) Navigate your browser to http://elbo-spice.cs.tau.ac.il/shiny/burrito/.

3) Upload the file **cohort1_16S_gg.txt** as the "Taxonomic abundances" file. 

4) Leave all the settings in the "Functional Data" box as the defaults. 

5) Upload the file **cohort1_16S_metadata.txt** as the "Sample grouping" file, and select one of the variables by which to group samples (recommended: "BV_Nugent_3"). 

6) Push "Assemble the BURRITO" and wait for the visualization to load. Push OK for any warnings that appear.

7) Explore the various ways to interact with the BURRITO visualization: Navigate up and down levels of the taxonomic and functional hierarchies by clicking on nodes in the two trees. View the specific abundance values by mousing over and/or clicking any portion of the visualization.

Follow these steps to visualize the second dataset, which demonstrates a different input format for BURRITO:

8) Refresh your browser to return to the BURRITO input page.

9) Upload **cohort2_qpcr.txt** as the "Taxonomic abundances" file.

10) Under Taxonomy, select "Custom taxonomy" and upload **cohort2_taxonomy_vaginal_burrito.txt** as the taxonomic hierarchy file.

11) Optionally, under Functional Data, select "Metagenome-based functional abundance table" and upload **cohort2_metagenomes.txt** as the functional comparison dataset. 

12) Under "Taxonomy-function linking method", select "Custom genomic content table" and upload **cohort2_qpcr_genomes.txt** as the genomic content file.

13) Upload **cohort2_metadata.txt** as the "Sample grouping" file and select one of the metadata variables by which to group samples (recommended: "BV").

14) Push "Assemble the BURRITO" and wait for the visualization to load. Push OK for any warnings that appear.


## Full-length Tutorial

### Overview and Objectives

BURRITO is a tool for exploratory data visualization and preliminary analysis. BURRITO can be used to answer questions like:

- What patterns exist in my taxonomic and functional data?
- How does taxonomic variation correspond with functional variation in my dataset?
- Which taxa are responsible for functional differences between my samples?
- What level of taxonomic and/or functional resolution is most interesting in my dataset?
- Are there significant differences in taxa and/or functions between sample groups?

BURRITO can also be used to generate static publication-quality visualizations.

In this tutorial, you will:

- Identify two different types of input data for BURRITO.
- Access the BURRITO web server, upload data, and generate a visualization.
- Try out the different modes of interaction to explore your data with BURRITO.
- Observe the various types of output files that can be produced by BURRITO.

### Installation
BURRITO is accessed via a web server and does not require installing any additional software.

### Input Data

If you have processed your own 16S rRNA data to a Greengenes OTU table, you can use it for this analysis. Otherwise, we will use two example datasets describing the vaginal microbiome. 

These datasets are from the following publication:

Srinivasan, S., Morgan, M.T., Fiedler, T.L., Djukovic, D., Hoffman, N.G., Raftery, D., Marrazzo, J.M., and Fredricks, D.N. (2015). Metabolic Signatures of Bacterial Vaginosis. MBio 6, e00204-15.

Download the example data by clicking and unzipping [this file](https://cnoecker.github.io/TAUMicrobiomeWorkshop2019/data/BurritoData.zip). 

The example data includes two different datasets, describing vaginal microbiome samples from 2 cohorts of women with and without Bacterial Vaginosis (BV).

- Cohort 1 (files **cohort1_16S_gg.txt**, **cohort1_16S_metadata.txt**): 16S rRNA data (n=30)

- Cohort 2 (files **cohort2_qpcr.txt**, **cohort2_qpcr_genomes.txt**, **cohort2_metagenomes.txt**, **cohort2_metadata.txt**, **cohort2_taxonomy_vaginal_burrito.txt**): 16S qPCR measurements for 14 common vaginal taxa, plus KEGG annotations of the genomic content of those taxa, plus metagenomic data with functional KO annotations (n=39)

Visualizing these datasets will demonstrate two of the three main input data options for BURRITO. The three options are:

1) **Greengenes OTU table (Cohort 1):** A closed-reference Greengenes 97% OTU table. By default, BURRITO uses functional content pre-computed by PICRUSt1 to calculate estimated functional abundances from Greengenes OTUs. 

2) **Taxonomic abundances plus genomic content (Cohort 2):** Taxonomic abundances in any format plus a custom genomic content table: In this case, BURRITO calculates estimated functional abundances and their attribution to different taxa using the genomic information provided. For this dataset, KEGG annotations associated with each species were downloaded from the Integrated Microbial Genomes database. The output of PICRUSt2 can also be used as input to BURRITO. 

3) **Function attribution table (not shown here):** BURRITO can also accept what it calls a "functional attribution table", which is a contribution table in the format produced by PICRUSt, describing the abundance of each gene (KEGG ortholog), from each taxon, in each sample. A stratified functional abundance table as produced by the Humann2 software could also be used for this purpose, with some re-formatting. 

See the documentation for more examples of each possible input format.

### Using your own data

If you would like to use your own 16S rRNA dataset, for this tutorial you should process it using qiime2 to obtain a closed-reference Greengenes OTU table. 

To do so, run the following commands in your qiime2 environment (which were also provided in the pre-workshop instructions). These assume that you have files named "rep-seqs.qza" and "table.qza" containing sequences and their abundances across samples. 

```{bash, eval = F}
# Import Greengenes reference OTU sequences
qiime tools import \
  --input-path 97_otus.fasta \
  --output-path 97_otus.qza \
  --type 'FeatureData[Sequence]'

## Here table.qza and rep-seqs.qza are the processed outputs from your Deblur or DADA2 analysis
qiime vsearch cluster-features-closed-reference \
  --i-table table.qza \
  --i-sequences rep-seqs.qza \
  --i-reference-sequences 97_otus.qza \
  --p-perc-identity 0.97 \
  --o-clustered-table table-cr-97.qza \
  --o-clustered-sequences rep-seqs-cr-97.qza \
  --o-unmatched-sequences unmatched-cr-97.qza

## Export the table from qiime, and convert the biom table to a tab-delimited table  
qiime tools export \
  --input-path table-cr-97.qza \
  --output-path greengenes-cr-feature-table

biom convert -i greengenes-cr-feature-table/feature-table.biom -o otu_table.txt --to-tsv

```

The **otu_table.txt** output of this code can be provided as the taxonomic abundances file to BURRITO.

#### Advanced input options

The input screen for BURRITO includes many additional options to customize your visualization for different types of data. To learn more about these, visit the BURRITO documentation linked at the top of this document.

### Using BURRITO
Navigate your web browser to the BURRITO website: http://elbo-spice.cs.tau.ac.il/shiny/burrito/ . You should see a panel of data input options. 

For dataset 1, BURRITO will use PICRUSt version one's pre-computed gene content for Greengenes OTUs. To view dataset 1: 

1) Upload **cohort1_16S_gg.txt** as the "Taxonomic abundances" file. 
2) Leave all the settings in the "Functional Data" box as the defaults. 
3) Upload **cohort1_16S_metadata.txt** as the "Sample grouping" file, and select one of the variables by which to group samples (recommended: "BV_Nugent_3"). 
4) Push "Assemble the BURRITO" and wait for the visualization to load. Push OK for any warnings that appear.

For dataset 2, we will provide a file to BURRITO specifying the KO gene content of each of the measured taxa in our dataset. To view dataset 2:

1) Upload **cohort2_qpcr.txt** as the "Taxonomic abundances" file.
2) Under Taxonomy, select "Custom taxonomy" and upload **cohort2_taxonomy_vaginal_burrito.txt** as the taxonomic hierarchy file.
3) Optionally, under Functional Data, select "Metagenome-based functional abundance table" and upload **cohort2_metagenomes.txt** as the functional comparison dataset. 
4) Under "Taxonomy-function linking method", select "Custom genomic content table" and upload **cohort2_qpcr_genomes.txt** as the genomic content file.
5) Upload **cohort2_metadata.txt** as the "Sample grouping" file and select one of the metadata variables by which to group samples (recommended: "BV").
6) Push "Assemble the BURRITO" and wait for the visualization to load. Push OK for any warnings that appear.

#### Interactions

The two bar plots display the taxonomic and functional abundances in the dataset. Navigate up and down levels of the taxonomic and functional hierarchies by clicking on nodes in the two trees. The size of each node in the tree corresponds with its average abundance, and the width of the connecting edges represents the average contribution of each taxon to each function. You can view the specific abundance values by mousing over and/or clicking any portion of the visualization. Clicking on any portion of the bar plots highlights all data linked to the selected feature (this also allows you to export a version of the plot with a specific feature highlighted).

Some questions to explore as you interact with BURRITO:

- Which taxa and functions are most variable across samples?
- Which taxa appear to be responsible for specific functional differences?
- At what levels of taxonomic and functions do you see interesting variation in samples?
- For example cohort 2, what differences can be seen between the taxa-based functional profile and the metagenomic functional profile?

#### Output

Click the settings flag in the upper left corner to display a menu. You can download SVG or PNG versions of the current state of the visualization as a whole or of individual plot components. You can also download the underlying processed data. If you grouped your samples into two groups, BURRITO will also perform some preliminary statistical analyses on differential abundances of the displayed features and make the results available for download. If you visualized a Greengenes OTU table, BURRITO also provides data on the NSTI metric of the quality of the PICRUSt predictions.

### Limitations to keep in mind

- Displaying and interacting with large datasets can take a long time. Consider summarizing to a higher level of resolution, filtering rare and low-abundance taxa, and/or viewing a subset of samples at a time. If the connection to the server is repeatedly lost, there may be too many people trying to access it at once.

- The statistical tests performed by BURRITO are exploratory analyses and may or may not be appropriate for your study. 

- The taxonomic and functional hierarchies displayed are only one possible way of grouping these features - for example, taxonomic distance does not necessarily correspond perfectly with phylogenetic distance. 

